larpandoracontent/LArObjects/LArDiscreteProbabilityVector.cc

```diff
@@ -0,0 +1,240 @@
+
+DiscreteProbabilityVector::DiscreteProbabilityData DiscreteProbabilityVector::RandomiseDiscreteProbabilityData(
+   DiscreteProbabilityVector const &discreteProbabilityVector, std::mt19937 &randomNumberGenerator) const
+{
+    DiscreteProbabilityData randomisedProbabilityData;
+
+    std::vector<size_t> randomisedElements(discreteProbabilityVector.GetSize());
+    std::iota (std::begin(randomisedElements), std::end(randomisedElements), 0);
+    std::shuffle(std::begin(randomisedElements), std::end(randomisedElements), randomNumberGenerator);
+
+    float xPos(discreteProbabilityVector.GetX(0));
+    float cumulativeProbability(0.f);
+    for (size_t iElement = 0; iElement < discreteProbabilityVector.GetSize(); ++iElement)
+    {
+        const size_t randomElementIndex(randomisedElements.at(iElement));
+        const float deltaX(discreteProbabilityVector.GetWidth(randomElementIndex));
+        const float probabilityDensity(discreteProbabilityVector.GetProbabilityDensity(randomElementIndex));
+        cumulativeProbability+=probabilityDensity*(m_useWidths ? deltaX : 1.f);
+        randomisedProbabilityData.emplace_back(DiscreteProbabilityVector::DiscreteProbabilityDatum(xPos, probabilityDensity, 
+            cumulativeProbability, deltaX));
+        xPos+=deltaX;
+    }
+
+    return randomisedProbabilityData;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+template <typename TX, typename TY>
+bool DiscreteProbabilityVector::SortInputDataByX(InputDatum<TX, TY> lhs, InputDatum<TX, TY> rhs)
+{
+    float deltaX(static_cast<float>(rhs.first) - static_cast<float>(lhs.first));
+    if (std::fabs(deltaX) < std::numeric_limits<float>::epsilon())
+        return (lhs.second < rhs.second);
+
+    return (lhs.first < rhs.first);
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+template <typename TX, typename TY>
+float DiscreteProbabilityVector::CalculateNormalisation(InputData<TX, TY> const &inputData) const
+{
+    if (2 > inputData.size())
+        throw pandora::StatusCodeException(pandora::STATUS_CODE_INVALID_PARAMETER);
+
+    float normalisation(0.f);
+
+    for (size_t iDatum = 0; iDatum < inputData.size()-1; ++iDatum)
+    {
+        const float deltaX(static_cast<float>(inputData.at(iDatum+1).first) - static_cast<float>(inputData.at(iDatum).first));
+        const float y(static_cast<float>(inputData.at(iDatum).second));
+        normalisation += y*(m_useWidths ? deltaX : 1.f);
+    }
+    const float deltaX(m_xUpperBound - static_cast<float>(inputData.back().first));
+    const float y(static_cast<float>(inputData.back().second));
+    normalisation += y*(m_useWidths ? deltaX : 1.f);
+
+    return normalisation;
+}
```

Comment:
Is there an easy way to avoid calculating `deltaX` when `m_useWidths = false`? Maybe anything you could try would have too many checks to make it faster / worthwhile.

Reply:
Maybe move the `?` part into the `deltaX` assignments?

Reply:
I deliberately left the width in place, regardless of whether they are used or not.  I figured it might still be useful for a user to know how far apart the bins are, even if the normalisation didn't factor in the width in the calculation.
What do you think?

Reply:
Done!